backend/pikpak/pikpak.go

```diff
@@ -0,0 +1,1749 @@
+		client:  client,
+		srv:     rest.NewClient(client).SetRoot(rootURL).SetErrorHandler(errorHandler),
+		pacer:   fs.NewPacer(ctx, pacer.NewDefault(pacer.MinSleep(minSleep), pacer.MaxSleep(maxSleep), pacer.DecayConstant(decayConstant))),
+	}
+	f.features = (&fs.Features{
+		ReadMimeType:            true, // can read the mime type of objects
+		CanHaveEmptyDirectories: true, // can have empty directories
+	}).Fill(ctx, f)
+
+	// Check if current token is valid and re-authorize if necessary.
+	// It is somehow redundant since it is covered in shouldRetry().
+	// However it is a preemptive measure to avoid possible retries while validating a token.
+	if token, err := oauthutil.GetToken(name, m); err == nil && !token.Valid() {
+		fs.Debugf(f, "Token is invalid. Trying to get a new one using username/password...")
+		if err := f.doAuthorize(ctx); err != nil {
+			return nil, fserrors.FatalError(err)
+		}
+	}
+
+	return f, nil
+}
+
+// NewFs constructs an Fs from the path, container:path
+func NewFs(ctx context.Context, name, path string, m configmap.Mapper) (fs.Fs, error) {
+	f, err := newFs(ctx, name, path, m)
+	if err != nil {
+		return nil, err
+	}
+
+	// Set the root folder ID
+	if f.opt.RootFolderID != "" {
+		// use root_folder ID if set
+		f.rootFolderID = f.opt.RootFolderID
+	} else {
+		// pseudo-root
+		f.rootFolderID = "root"
+	}
+
+	f.dirCache = dircache.New(f.root, f.rootFolderID, f)
+
+	// Find the current root
+	err = f.dirCache.FindRoot(ctx, false)
+	if err != nil {
+		// Assume it is a file
+		newRoot, remote := dircache.SplitPath(f.root)
+		tempF := *f
+		tempF.dirCache = dircache.New(newRoot, f.rootFolderID, &tempF)
+		tempF.root = newRoot
+		// Make new Fs which is the parent
+		err = tempF.dirCache.FindRoot(ctx, false)
+		if err != nil {
+			// No root so return old f
+			return f, nil
+		}
+		_, err := tempF.NewObject(ctx, remote)
+		if err != nil {
+			if err == fs.ErrorObjectNotFound {
+				// File doesn't exist so return old f
+				return f, nil
+			}
+			return nil, err
+		}
+		f.features.Fill(ctx, &tempF)
+		// XXX: update the old f here instead of returning tempF, since
+		// `features` were already filled with functions having *f as a receiver.
+		// See https://github.com/rclone/rclone/issues/2182
+		f.dirCache = tempF.dirCache
+		f.root = tempF.root
+		// return an error with an fs which points to the parent
+		return f, fs.ErrorIsFile
+	}
+	return f, nil
+}
+
+// readMetaDataForPath reads the metadata from the path
+func (f *Fs) readMetaDataForPath(ctx context.Context, path string) (info *api.File, err error) {
+	leaf, dirID, err := f.dirCache.FindPath(ctx, path, false)
+	if err != nil {
+		if err == fs.ErrorDirNotFound {
+			return nil, fs.ErrorObjectNotFound
+		}
+		return nil, err
+	}
+
+	// checking whether fileObj with name of leaf exists in dirID
+	trashed := "false"
+	if f.opt.TrashedOnly {
+		trashed = "true"
+	}
+	found, err := f.listAll(ctx, dirID, api.KindOfFile, trashed, func(item *api.File) bool {
```

Comment:
If there is a quicker way of finding a given file than listing the parent, then this `readMetaDataForPath` should implement it. It makes quite a difference to performance if such a call is available.